README.md
Chez Betty
==========

[![Build Status](https://travis-ci.org/um-cseg/chez-betty.svg?branch=master)](https://travis-ci.org/um-cseg/chez-betty)
[![Code Climate](https://codeclimate.com/github/um-cseg/chez-betty/badges/gpa.svg)](https://codeclimate.com/github/um-cseg/chez-betty)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/um-cseg/chez-betty/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/um-cseg/chez-betty/?branch=master)

Chez Betty is a mini, cooperative food store where users can deposit money to their account
and use it to purchase food and other items. The system is linked
to the UMich ldap server so users can swipe their M-Card to login.

UMich version is an homage to UCSD's [Chez Bob](http://chezbob.ucsd.edu/).
Chez Betty runs [here](http://chezbetty.store).

Chez Betty is written as a Python web app designed to run on a server with
a browser based user interface. It supports cash and credit card deposits. The
user interface expects a card swiper and barcode scanner to be attached to
the machine. See the list at the end of this README for products we use.
To test / demo the system without hardware, turn on "Demo mode" in the admin
interface.

Installation
------------

The Chez Betty application is developed using Python Pyramid. To get a
development environment set up, do the following:

1. Clone this repository

        git clone https://github.com/um-cseg/chez-betty.git

2. Install system dependencies:

        # Ubuntu 15.10
        sudo apt-get install postgresql postgresql-server-dev-9.4 libjpeg-dev
        # Ubuntu 16.10
        sudo apt-get install postgresql postgresql-server-dev-9.5 libjpeg-dev

1. Setup virtualenv with Python3:

        cd chez-betty

        # Python cannot decide on how it wants to do this. So
        # we have multiple iterations that we have tried.
        ## Way back:
        # virtualenv .
        ## Python3.3:
        # python3 -m venv .
        # ^ *not* pyvenv ., that installed python2 in my venv for w/e reason
        ## Newest (only do this with Python 3.4+):
        # pyvenv-3.5 venv
        ## Python 3.6 changes it again!
        python3.6 -m venv

        source venv/bin/activate

2. Install the dependencies:

        # Note, this step *must* be run before setup.py. The latter will use
        # easy_install to install requirements instead of pip, which will
        # break some packages (e.g. stripe). The argument should be the root
        # directory of the project.
        pip install -e .

2. Setup the development environment

        python setup.py develop

3. Update `development.ini` to set config information and passwords.

        cp development.ini.example development.ini
        [edit development.ini]

3. Setup the database:

        python chezbetty/initializedb.py development.ini

4. Install bower and bower-install

        # Note: Distro copies of node tend to fall behind really quickly.
        # You are better off using a copy of node/npm installed directly:
        # http://nodejs.org/
        # https://github.com/npm/npm
        sudo npm install bower bower-installer -g

5. Get all css/js dependencies

        bower-installer

4. Run the webserver:

        pserve development.ini
        # n.b. pserve will be in your path if your virtualenv is active


### LDAP

TODO: add information about how to setup LDAP connection.

### Bitcoin

TODO: add info about how to setup bitcoin/coinbase.

Usage
-----

View the client interface using

    http://127.0.0.1:6543

View the admin interface using

    http://127.0.0.1:6543/admin
    username: admin
    password: chezbettyadmin

Tools
-----

Chez Betty currently runs using:

- Intel Nuc with case - Running Ubuntu.
- [Magnetic Strip Reader](https://www.cdw.com/shop/products/MagTek-SureSwipe-Reader-USB-HID-Keyboard-Interface-magnetic-card-reader/1140626.aspx) - Acts as a keyboard input.
- [Barcode Scanner](https://www.cdw.com/shop/products/Motorola-LS2208-barcode-scanner-scanner-and-USB-cable-included/3021140.aspx) - Also keyboard input.
- [Touchscreen Monitor](http://www.amazon.com/ViewSonic-TD2220-22-Inch-LED-Lit-Display/dp/B009F1IKFC) - Acts as a mouse.


HISTORY.md
History of Chez Betty
=====================

- April 12, 2019: Betty's second official opening, almost five years after
the first. Betty is back!!
- April 8, 2019: Betty has a soft opening. The shelves only contain Soylent,
Starbucks, and BBQ chips, but things are running again!
- April 3, 2019: After nearly two years, Chez Betty is approved to re-open!
- April 1, 2019: The freezer is removed because Betty no longer carries any
frozen food.
- Summer, 2017: Betty stops restocking and is temporarily shut down having run
afoul of University regulations on student run food operations. The maintainers
of Betty begin the process to meet University policies.
- April, 2017: The University notifies Betty that it does not meet University
requirements and must close. Gracious supporters launch a call and email
campaign to try to save Betty. The University relents and permits Betty
to apply to re-open after meeting the requirements.
- January 3, 2017: Added new shelves. Both for drink storage and for
more food display.
- June 16, 2016: Tried to hire a moving company to pick up our
Costco order, and have Costco put our order on pallets. We did get the
food, but it was a bit of an ordeal.
- June 9, 2016: Added a 42"x18"x54" inch shelf to betty to hold bins
full of chips and other bagged food.
- May 14, 2016: Added a bill acceptor to Betty. This should prevent
most of the loss from cash deposits being entered but the money
not being in the cash box.
- May 14, 2016: Replaced the thermostat in the True GDM-72 refrigerator.
The old one failed, causing the fridge to get too warm, then failed
the other way causing a lot of drinks to freeze. All told, it cost
Betty about $250.
- Dec 9, 2015: Betty ditches the BeagleBone Black and moves to an Intel
Nuc. Hopefully that will be more reliable and responsive.
- Sep 20, 2015: Betty restocks start requiring renting a U-Haul. With purchases at
Costco over $6k, this is the only way to move that much inventory.
- May 20, 2015: Betty bought the triple-wide fridge for $1200. Took 6 or so people
to get it into 2909, but actually cold drinks made it worth it.
- March 2015: Betty bought a freezer. After placing an ice cream order
with the Nestle rep at Kroger, Betty needed more space for frozen items.
Even with the expanded capacity, ice cream remains one of the fastest
sellers.
- July 10, 2014: One of the coolers from the Foo Bar is brought up to the
grad lounge and becomes cold storage for Chez Betty drinks.
- July 8, 2014: Chez Betty is installed in 2817 BBB. After a Costco run, adding
the first batch of items, stocking the shelves, and getting Internet to
the touchscreen and Beagle Bone Black, Betty is alive at 11 PM.
- June 27-29, 2014: 4901 BBB becomes the Chez Betty development office.
- June 20, 2014: Zakir and Brad buy all of the hardware that will become Betty.
- May 2014: After months of procrastination, Brad, Zakir, and Pat decided to set
aside a weekend and build Chez Betty. Pat creates an account with ITS for LDAP
access, but nothing else happens for a while.
- Summer 2013: A video chat with Tristan Halvorson at UCSD reveals that Chez
Betty is definitely worth doing. However, it is clear we are going to have to
build our own system.
- Summer 2013: Brad Campbell, Meghan Clark, Zakir Durumeric, Pat Pannuto decide
"Hey, if UCSD has Chez Bob, we should have Chez Betty."


export_transactions.py
#!./venv/bin/python

# https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/commandline.html#writing-a-script
from pyramid.paster import bootstrap

# Betty imports
from chezbetty.models.event import Event
from chezbetty.models.transaction import Transaction
from chezbetty.models.user import User

# General Python
import sys

import arrow
import xlsxwriter




year = int(sys.argv[1])
assert(year >= 2014)

START = arrow.get(year, 7, 1)
END = arrow.get(year+1, 7, 1)
#END = arrow.get(2014, 8, 1)

workbook = xlsxwriter.Workbook('records-{}.xlsx'.format(year), {'remove_timezone': True})

# Add a bold format to use to highlight cells.
bold = workbook.add_format({'bold': True})

# Add a number format for cells with money.
money_format = workbook.add_format({'num_format': '$#,##0.00'})

# Add an Excel date format.
date_format = workbook.add_format({'num_format': 'mmmm d yyyy hh:mm'})

sheet_overview = workbook.add_worksheet('Overview')
sheet_overview.write('A1', 'Type', bold)
sheet_overview.write('B1', 'Total', bold)
sheet_overview.write('C1', 'Explanation', bold)
sheet_overview.set_column(0, 1, 15)
sheet_overview.set_column(2, 2, 100)
overview_row = 1

sheet_banktransfer = workbook.add_worksheet('CashDeposits')
sheet_banktransfer.write('A1', 'Id', bold)
sheet_banktransfer.write('B1', 'Date', bold)
sheet_banktransfer.set_column(1, 1, 20)
sheet_banktransfer.write('C1', 'Amount', bold)
sheet_banktransfer.write('D1', 'Type', bold)
# When the "safe" was introduced to Betty software
BANKTRANSFER_CUTOVER_TIME = arrow.get(2016, 4, 14)
banktransfer_row = 1

sheet_charitable = workbook.add_worksheet('charitable')
sheet_charitable.write('A1', 'Id', bold)
sheet_charitable.write('B1', 'Date', bold)
sheet_charitable.set_column(1, 1, 20)
sheet_charitable.write('C1', 'Amount', bold)
sheet_charitable.write('D1', 'Type', bold)

cell = {}

with bootstrap('development.ini') as env:
    for transaction_type, explanation in (
            ('purchase', 'Money from user balance. Does not represent cash received. Does represent inventory removed'),
            ('deposit', 'Actual money from user into Betty.'),
            ('cashdeposit', 'Money *reported by users* into Betty'),
            ('ccdeposit', 'Actual money from user into Betty (via Credit Card).'),
            ('btcdeposit', 'Actual money from user into Betty (via BitCoin).'),
            ('adjustment', 'Catch-all for corrections/fixes/manual intervention'),
            ('restock', 'Actual money paid out to suppliers for Betty inventory'),
            ('inventory', 'Value of inventory lost when physical inventory counted'),
            ('emptycashbox', 'Money removed from the cashbox'),
            ('emptysafe', 'Money removed from the safe'),
            ('lost', 'Money expected in the cashbox/safe that was missing'),
            ('found', 'More money was in the cashbox/safe than expected'),
            ('donation', 'Money given to Betty'),
            ('withdrawal', 'Money removed from Betty for expenses'),
            ('reimbursement', 'Money removed from bank account for reimbursement'),
            ):
        transactions = Transaction.all(start=START, end=END, trans_type=transaction_type)

        sheet = workbook.add_worksheet(transaction_type)
        sheet.write('A1', 'Id', bold)
        sheet.write('B1', 'Date', bold)
        sheet.set_column(1, 1, 20)
        sheet.write('C1', 'Amount', bold)
        if transaction_type == 'purchase':
            sheet.write('D1', 'Wholesale', bold)
        else:
            sheet.write('D1', 'Notes', bold)
            sheet.set_column(3, 3, 100)

        row = 1
        for t in reversed(transactions):
            # Ignore initial capital "donation"s
            if t.id == 1485 or t.id == 2053:
                continue

            sheet.write_number(row, 0, t.id)
            sheet.write_datetime(row, 1, t.event.timestamp.datetime, date_format)
            sheet.write_number(row, 2, t.amount, money_format)
            if transaction_type == 'purchase':
                wholesale = 0
                for line_item in t.subtransactions:
                    wholesale += line_item.wholesale
                sheet.write_number(row, 3, wholesale, money_format)
            else:
                sheet.write(row, 3, t.event.notes)
            row += 1

            if transaction_type == 'emptycashbox' and t.event.timestamp < BANKTRANSFER_CUTOVER_TIME:
                sheet_banktransfer.write_number(banktransfer_row, 0, t.id)
                sheet_banktransfer.write_datetime(banktransfer_row, 1, t.event.timestamp.datetime, date_format)
                sheet_banktransfer.write_number(banktransfer_row, 2, t.amount, money_format)
                sheet_banktransfer.write(banktransfer_row, 3, "cashbox")
                banktransfer_row += 1
            if transaction_type == 'emptysafe':
                sheet_banktransfer.write_number(banktransfer_row, 0, t.id)
                sheet_banktransfer.write_datetime(banktransfer_row, 1, t.event.timestamp.datetime, date_format)
                sheet_banktransfer.write_number(banktransfer_row, 2, t.amount, money_format)
                sheet_banktransfer.write(banktransfer_row, 3, "safe")
                banktransfer_row += 1

        cell[transaction_type] = 'B{}'.format(overview_row+1)
        sheet_overview.write(overview_row, 0, transaction_type)
        sheet_overview.write_formula(overview_row, 1,
                '=SUM({}!C:C)'.format(transaction_type), money_format)
        sheet_overview.write(overview_row, 2, explanation)
        overview_row += 1
        if transaction_type == 'purchase':
            cell['purchase_wholesale'] = 'B{}'.format(overview_row+1)
            sheet_overview.write(overview_row, 0, 'purchase_wholesale')
            sheet_overview.write_formula(overview_row, 1,
                    '=SUM({}!D:D)'.format(transaction_type), money_format)
            sheet_overview.write(overview_row, 2, 'wholesale cost of purchases')
            overview_row += 1

cell['CashDeposits'] = 'B{}'.format(overview_row+1)
sheet_overview.write(overview_row, 0, 'CashDeposits')
sheet_overview.write_formula(overview_row, 1, '=SUM(CashDeposits!C:C)', money_format)
sheet_overview.write(overview_row, 2, 'Actual money put into the bank')
overview_row += 1

cell['charitable'] = 'B{}'.format(overview_row+1)
sheet_overview.write(overview_row, 0, 'charitable')
sheet_overview.write_formula(overview_row, 1, '=SUM(charitable!C:C)', money_format)
sheet_overview.write(overview_row, 2, 'Donations from Betty to CSEG')
overview_row += 1

overview_row += 3
sheet_overview.write(overview_row, 0, 'Actual Income')
sheet_overview.write_formula(overview_row, 1,
        '=SUM({},{},{},{})'.format(cell['ccdeposit'], cell['btcdeposit'], cell['CashDeposits'], cell['donation']), money_format)
sheet_overview.write(overview_row, 2, 'Money that hit the bank account (CC and BTC deposits, plus actual cash in cashbox/safe, plus donations)')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Accrued Income')
sheet_overview.write_formula(overview_row, 1,
        '=SUM({},{},{},{})'.format(cell['cashdeposit'], cell['ccdeposit'], cell['btcdeposit'], cell['donation']), money_format)
sheet_overview.write(overview_row, 2, 'Money that should hit the bank account (User-reported cash, CC and BTC deposits, plus donations)')
overview_row += 1
sheet_overview.write(overview_row, 0, 'User Liabilities')
sheet_overview.write_formula(overview_row, 1,
        '=SUM({},{},{})-{}'.format(cell['cashdeposit'], cell['ccdeposit'], cell['btcdeposit'], cell['purchase']), money_format)
sheet_overview.write(overview_row, 2, 'Money in user accounts not yet spent (Cash/CC/BTC deposits - purchases)')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Actual Outflow')
sheet_overview.write_formula(overview_row, 1,
        '=SUM({},{},{})'.format(cell['adjustment'], cell['restock'], cell['withdrawal']), money_format)
sheet_overview.write(overview_row, 2, 'Money that left the bank account (adjustment + restock + withdrawal)')


########################################
## Begin Form 1125-A
overview_row += 3
sheet_overview.write(overview_row, 0, 'Form 1125-A', bold)
overview_row += 1
sum_start = overview_row
cell['inventory_beginning'] = 'B{}'.format(overview_row+1)
sheet_overview.write(overview_row, 0, 'Box 1')
sheet_overview.write(overview_row, 2, 'Get from prior year Line 7')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 2')
sheet_overview.write_formula(overview_row, 1, '='+cell['restock'], money_format)
sheet_overview.write(overview_row, 2, 'Purchases')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 3')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Cost of Labor')
overview_row += 1
# Resellers with less than $10Million average over the last three years
sheet_overview.write(overview_row, 0, 'Box 4')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Section 263A costs')
overview_row += 1
sum_end = overview_row
sheet_overview.write(overview_row, 0, 'Box 5')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Other costs')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 6')
sheet_overview.write_formula(overview_row, 1,
        '=SUM(B{}:B{})'.format(sum_start+1,sum_end+1), money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 7')
sheet_overview.write_formula(overview_row, 1,
        '=' + cell['inventory_beginning'] + '+' + cell['restock'] + '-' + cell['purchase_wholesale'] + '-' + cell['inventory'], money_format)
sheet_overview.write(overview_row, 2, 'Inventory from the year is sum restocks less wholesale purchase values less inventory losses')
overview_row += 1
cell['COGS'] = 'B{}'.format(overview_row+1)
sheet_overview.write(overview_row, 0, 'Box 8')
sheet_overview.write_formula(overview_row, 1,
        '=B{}-B{}'.format(overview_row-2+1,overview_row-1+1), money_format)
sheet_overview.write(overview_row, 2, 'Cost of goods sold')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 9')
sheet_overview.write(overview_row, 2, 'Check 9a(i) [Cost], 9b-d blank, 9e and 9f check No')



########################################
## Begin Form 1120
overview_row += 3
sheet_overview.write(overview_row, 0, 'Form 1120', bold)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 1a')
sheet_overview.write_formula(overview_row, 1, '='+cell['purchase'], money_format)
sheet_overview.write(overview_row, 2, 'Gross receipts or sales')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 1b')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 1c')
sheet_overview.write_formula(overview_row, 1, '='+cell['purchase'], money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 2')
sheet_overview.write_formula(overview_row, 1, '='+cell['COGS'], money_format)
sheet_overview.write(overview_row, 2, 'Cost of goods sold')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 3')
sum_start = overview_row
sheet_overview.write_formula(overview_row, 1,
        '=B{}-B{}'.format(overview_row-2+1,overview_row-1+1), money_format)
sheet_overview.write(overview_row, 2, 'Gross profit')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 4')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 5')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 6')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 7')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 8')
sheet_overview.write_number(overview_row, 1, 0, money_format)
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 9')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Form 4797: Sales of Business Property')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 10')
sum_end = overview_row
sheet_overview.write_formula(overview_row, 1,
        '={}+{}'.format(cell['found'], cell['donation']), money_format)
sheet_overview.write(overview_row, 2, 'Other income. Attach Schedule. DO THIS')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 11')
cell['total_income'] = 'B{}'.format(overview_row+1)
sheet_overview.write_formula(overview_row, 1,
        '=SUM(B{}:B{})'.format(sum_start+1,sum_end+1), money_format)
sheet_overview.write(overview_row, 2, 'Total income')

overview_row += 1
sum_start = overview_row
sheet_overview.write(overview_row, 0, 'Box 12')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Compensation of officers')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 13')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Salaries and wages')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 14')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Repairs and maintenance. No machinery/big things, careful do not double-count')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 15')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'We could do this, but then would also have to do the flip side (absorbed user balances). Easier to leave all outstanding.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 16')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Rents. (Includes vehicle rentals)')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 17')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Taxes and licenses. Do not include sales tax.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 18')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Interest (from debts).')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 19')
#Need to calculate later with 10% cap
#sheet_overview.write_formula(overview_row, 1,
#        '={}'.format(cell['charitable']), money_format)
sheet_overview.write(overview_row, 2, 'Charitable contributions.')
charitable_row = overview_row
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 20')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Depreciation of assets.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 21')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Depletion (of natural deposits).')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 22')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Advertising.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 23')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Pension/profit-sharing/etc.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 24')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Employee Benefit Programs.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 25')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Domestic production deduction.')
overview_row += 1
sum_end = overview_row
sheet_overview.write(overview_row, 0, 'Box 26')
sheet_overview.write_formula(overview_row, 1,
        '={}+{}'.format(cell['lost'], cell['withdrawal']), money_format)
sheet_overview.write(overview_row, 2, 'Other deductions (attach statement) DO THIS.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 27')
cell['total_deductions'] = 'B{}'.format(overview_row+1)
sheet_overview.write_formula(overview_row, 1,
        '=SUM(B{}:B{})'.format(sum_start+1,sum_end+1), money_format)
sheet_overview.write(overview_row, 2, 'Total deductions')
sheet_overview.write_formula(overview_row, 3,
        '=SUM(B{}:B{}) + SUM(B{}:B{})'.format(
            sum_start+1,charitable_row-1+1,
            charitable_row+1+1,sum_end+1), money_format)
cell['total_deductions_no_charitable'] = 'D{}'.format(overview_row+1)
sheet_overview.write(overview_row, 4, 'Deductions except charitable')

overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 28')
sheet_overview.write_formula(overview_row, 1,
        '=' + cell['total_income'] + '-' + cell['total_deductions'], money_format)
sheet_overview.write(overview_row, 2, 'Taxable income before NOL')
sheet_overview.write_formula(overview_row, 3,
        '=' + cell['total_income'] + '-' + cell['total_deductions_no_charitable'], money_format)
sheet_overview.write(overview_row, 4, 'Taxable income except charitable')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 29abc')
sheet_overview.write_number(overview_row, 1, 0, money_format)
sheet_overview.write(overview_row, 2, 'Net operating losses (NOL).')
sheet_overview.write_formula(overview_row, 3,
        '=0.1*D{}'.format(overview_row-1+1), money_format)
sheet_overview.write(overview_row, 4, '10% taxable income except charitable')
cell['charitable_cap'] = 'D{}'.format(overview_row+1)
#Late-calculate charitable
sheet_overview.write_formula(charitable_row, 1,
        '=MIN({},{})'.format(cell['charitable'],cell['charitable_cap']), money_format)

overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 30')
sheet_overview.write_formula(overview_row, 1,
        '=B{}-B{}'.format(overview_row-2+1,overview_row-1+1), money_format)
sheet_overview.write(overview_row, 2, 'Taxable income.')
overview_row += 1
sheet_overview.write(overview_row, 0, 'Box 31')
sheet_overview.write_formula(overview_row, 1,
        '=B{} * 0.15'.format(overview_row-1+1), money_format)
sheet_overview.write(overview_row, 2, 'CHECK INSTRUCTIONS / do J: 15% for <$50k and no deductions assumed')


workbook.close()
